refiner-lib/src/refiner_inner.rs

```diff
@@ -309,13 +324,133 @@ struct BuiltTransaction {
+        }
+        Some(ExecutionStatusView::SuccessReceiptId(result)) => {
+            // No need to check the transaction kind in this case because transactions that
+            // produce a SubmitResult as output do not produce a receipt id.
+            let bytes = result.0.to_vec();
+            Some(SubmitResult::new(
+                aurora_engine::parameters::TransactionStatus::Succeed(bytes),
+                21_000,
+                Vec::new(),
+            ))
+        }
+        None => None,
+    };
+
+    let engine_output = engine_outcome
+        .and_then(|x| x.maybe_result.as_ref().ok())
+        .and_then(Option::as_ref)
+        .map(|result| match result {
+            TransactionExecutionResult::Submit(result) => match result {
+                Ok(result) => result.clone(),
+                Err(err) => SubmitResult::new(
+                    aurora_engine::parameters::TransactionStatus::Revert(
+                        format!("{:?}", err.kind).into_bytes(),
+                    ),
+                    err.gas_used,
+                    Vec::new(),
+                ),
+            },
+            TransactionExecutionResult::DeployErc20(address) => SubmitResult::new(
+                aurora_engine::parameters::TransactionStatus::Succeed(address.as_bytes().to_vec()),
+                21_000,
+                Vec::new(),
+            ),
+            TransactionExecutionResult::Promise(p) => SubmitResult::new(
+                aurora_engine::parameters::TransactionStatus::Succeed(
+                    format!("{:?}", p).into_bytes(),
+                ),
+                21_000,
+                Vec::new(),
+            ),
+        });
+
+    match (near_output, engine_output) {
+        (Some(near_output), Some(engine_output)) => {
+            // We have a result from both sources, so we should compare them to
+            // make sure they match. Log a warning and use the Near output if they don't.
+            if near_output != engine_output {
+                tracing::warn!("Mismatch between Near and Engine outputs. The internal Engine instance may not have the correct state.");
+            }
+            Ok(near_output)
+        }
+        (None, Some(output)) => {
+            // No Near outcome to rely on, so we simply have to trust the Borealis Engine
+            // outcome without validation. This case happens for actions in a batch except
+            // for the last one (Near only records the outcome of the last action in a batch).
+            Ok(output)
+        }
+        (Some(output), None) => {
+            // No engine outcome to use, so can only rely on the NEAR output.
+            // This case could arise if the last action in a batch is an aurora-engine call
+            // where the Borealis Engine does not record an outcome (e.g. `ft_on_transfer`).
+            Ok(output)
+        }
+        (None, None) => {
+            // if there is no outcome from either source then use a default value
```

Comment:
IMO, not obvious what `21_000` is. I'd create a constant for this or at least modify the comment.
```suggestion
            // if there is no outcome from either source, then use a default value which is 21_000
```

Reply:
Good suggestion. Done in https://github.com/aurora-is-near/borealis-engine-lib/pull/35/commits/dc3531e89629ce1a91e0bcec2f5e8e79886dbbfd